### ISSUE TITLE
US-integration of RBC Model and Track model 

### ISSUE THREAD

**Issue by BaseliyosJacob** [openETCS/modeling#497]
@mairamou @JakobGartner @aprobst For the "proof of concept" of Utrecht - Amsterdam we need the integration of the RBC Model with the Utrecht - Amsterdam Track model

**Comment by JakobGartner** [comment]
This is in work as far as @mairamou and I are concerned and we hope to
commit it before Monday

Jakob



On 27.06.2015, at 08:59, Baseliyos Jacob <notifications@github.com> wrote:

@mairamou <https://github.com/mairamou> @JakobGartner
<https://github.com/JakobGartner> @aprobst <https://github.com/aprobst> For
the "proof of concept" of Utrecht - Amsterdam we need the integration of
the RBC Model with the Utrecht - Amsterdam Track model

—
Reply to this email directly or view it on GitHub
<https://github.com/openETCS/modeling/issues/497>.

**Comment by aprobst** [comment]
@mairamou, thank you for your provided work and excel sheet in #446. Taking a closer look to the exel document, some inconsistencies may exists in cell 9:

(9, LRBG_353_D_00431_0_M024, 24, P57, 353004310, 353, 431.0, M_ACK set to 0 (not required), actual setting not visible from JRU data, but as no T_TRAIN is provided this was concluded)

I think that this entry refers to the interaction from user story 7 in which the RBC sends its MA request and position report parameters to the train with message 24. If this is the case, then this message must be acknowledged and also contain packet 58.

I think @BaseliyosJacob can give us a clarification if needed.



Also I would like add that empty messages with NID_MESSAGE=24, which serve as a keep-alive message, do not need to be integraded in the track model. This is due to the fact that this kind of messages base on a triggering mechanism that depend on the T_NVCONTACT parameter.

Curently, such kind of messages are automatically sent if following condition is met:

(no radio based track train message to send) AND ( (current RBC time - time of last sent radio track train message) >= k * T_NVCONTACT ) , where k is a real number with 0 < k << 1



Last but not least, I have a question to @JakobGartner. In the project UtrechtAmsterdam_oETCS.etp, the operator FirstTest::Story00A_BTM exists. @JakobGartner, can you provide a similar operator like Story00A_RADIO with following signature:

function Story00A_RADIO(triggerID : int) returns  (radioTrackTrainMessage : TM::CompressedRadioMessage);

Thanks in advance!

**Comment by mairamou** [comment]
Hello Alexander, 

thnak you. 

I am very happy when I don't have to include the messages 24, as they
are a lot of just copy/paste work. 

I have removed them from the model with the exception of 

 		LRBG_353_D_00431_0_M024

where they also contain P57+P58 

If you want to generate this yourself, just don't trigger mine.. and I
can remove them later 

I have also set M_ACK to 1 for this message. 

Mairamou 

On 2015-06-28 22:34, Alexander Probst wrote: 

> @mairamou [1], thank you for your provided work and excel sheet in #446 [2]. Taking a closer look to the exel document, some inconsistencies may exists in cell 9: 
> 
> (9, LRBG_353_D_00431_0_M024, 24, P57, 353004310, 353, 431.0, M_ACK set to 0 (not required), actual setting not visible from JRU data, but as no T_TRAIN is provided this was concluded) 
> 
> I think that this entry refers to the interaction from user story 7 in which the RBC sends its MA request and position report parameters to the train with message 24. If this is the case, then this message must be acknowledged and also contain packet 58. 
> 
> I think @BaseliyosJacob [3] can give us a clarification if needed. 
> 
> Also I would like add that empty messages with NID_MESSAGE=24, which serve as a keep-alive message, do not need to be integraded in the track model. This is due to the fact that this kind of messages base on a triggering mechanism that depend on the T_NVCONTACT parameter. 
> 
> Curently, such kind of messages are automatically sent if following condition is met: 
> 
> (no radio based track train message to send) AND ( (current RBC time - time of last sent radio track train message) >= k * T_NVCONTACT ) , where k is a real number with 0 < k << 1 
> 
> Last but not least, I have a question to @JakobGartner [4]. In the project UtrechtAmsterdam_oETCS.etp, the operator FirstTest::Story00A_BTM exists. @JakobGartner [4], can you provide a similar operator like Story00A_RADIO with following signature: 
> 
> function Story00A_RADIO(triggerID : int) returns (radioTrackTrainMessage : TM::CompressedRadioMessage); 
> 
> Thanks in advance! 
> 
> --
> Reply to this email directly or view it on GitHub [5].

-- 
Mairamou Haman Adji
Student Trainee / Embedded Systems
---------------------------
LEA RAILERGY

 

Links:
------
[1] https://github.com/mairamou
[2] https://github.com/openETCS/modeling/issues/446
[3] https://github.com/BaseliyosJacob
[4] https://github.com/JakobGartner
[5]
https://github.com/openETCS/modeling/issues/497#issuecomment-116329799

**Comment by JakobGartner** [comment]
1. on the connection with the RBC, this is done. We have provided a comprehensive trigger mechanism which is document using a cross reference table which is in the project directory (and linked from the SCADE project)
2. on the connection with the OBU input channel, there is TODO:
- merging the balise and RBC streams using an input buffer. 
- as the Subset 26.8 messages are now merged into one stream, a (temporary) conversion function to the existing radio track to train header is needed. (additional variables may be added depending on the actual message. @BerndHekele : we need to merge the type TM_radio_messages;;M_TrackTrain_Radio_T with the existing Track to Train type.

Workload from TrackMessages/InfraLib team for the open issues: 0,5 days

**Comment by mairamou** [comment]
Hello Alexander, 

as for your question about Story00A that you asked to Jakob: 

I think it is time to slowly abandon Story00A and move to
AmsterdamUtrecht Level 2 track. 

There we have provided exactly the split that you propose. 

Story00A is not so nice because it is a bit fake, using a L1 structure. 

I am very sure that there are bugs in my test track, so we should start
testing it. 

Also many of the TrackMessages functions may still contain little
problems, as we had not enough time for testing it all. 

On 2015-06-28 22:34, Alexander Probst wrote: 

> @mairamou [1], thank you for your provided work and excel sheet in #446 [2]. Taking a closer look to the exel document, some inconsistencies may exists in cell 9: 
> 
> (9, LRBG_353_D_00431_0_M024, 24, P57, 353004310, 353, 431.0, M_ACK set to 0 (not required), actual setting not visible from JRU data, but as no T_TRAIN is provided this was concluded) 
> 
> I think that this entry refers to the interaction from user story 7 in which the RBC sends its MA request and position report parameters to the train with message 24. If this is the case, then this message must be acknowledged and also contain packet 58. 
> 
> I think @BaseliyosJacob [3] can give us a clarification if needed. 
> 
> Also I would like add that empty messages with NID_MESSAGE=24, which serve as a keep-alive message, do not need to be integraded in the track model. This is due to the fact that this kind of messages base on a triggering mechanism that depend on the T_NVCONTACT parameter. 
> 
> Curently, such kind of messages are automatically sent if following condition is met: 
> 
> (no radio based track train message to send) AND ( (current RBC time - time of last sent radio track train message) >= k * T_NVCONTACT ) , where k is a real number with 0 < k << 1 
> 
> Last but not least, I have a question to @JakobGartner [4]. In the project UtrechtAmsterdam_oETCS.etp, the operator FirstTest::Story00A_BTM exists. @JakobGartner [4], can you provide a similar operator like Story00A_RADIO with following signature: 
> 
> function Story00A_RADIO(triggerID : int) returns (radioTrackTrainMessage : TM::CompressedRadioMessage); 
> 
> Thanks in advance! 
> 
> --
> Reply to this email directly or view it on GitHub [5].

-- 
Mairamou Haman Adji
Student Trainee / Embedded Systems
---------------------------
LEA RAILERGY

 

Links:
------
[1] https://github.com/mairamou
[2] https://github.com/openETCS/modeling/issues/446
[3] https://github.com/BaseliyosJacob
[4] https://github.com/JakobGartner
[5]
https://github.com/openETCS/modeling/issues/497#issuecomment-116329799

**Comment by aprobst** [comment]
Good morning @mairamou! I agree with you.

From my point of view, I generate inside my project a trigger ID depending on LRBG and NID_LRBG. Now it would be of course fine if I could just import the UtrechtAmsterdam_oETCS.etp project into mine and call a provided operator from your side and pass this trigger value to it. As a result, your operator returns me a compressed radio message.

For example: let us assume that you created a whole level 2 track for the radio messages similar to Story00A and called the outer operator "Track_Radio". Then e.g. the following signature would satisfy my needs: 

function Track_Radio(triggerID : int) returns (radioTrackTrainMessage : TM::CompressedRadioMessage);

--> a trigger ID as an input and a compressed radio message as an output

Of course you can put the whole track including the Balise messages into one operator and call it e.g. "Track". As a result, you would use more input arguments and more outputs. For me it doesn't matter as long as the operator contains an input that takes a trigger ID and returns me a compressed radio message.

**Comment by mairamou** [comment]
Hello Alexander, 

the current track model works like this: 

- you send a trigger 

- the track model sends the corresponding message and if applicable,
packets as defined on the track. 

It can be integrated in two ways: 

1. "After" your model. You send the trigger, and the track model sends
the corresponding message/packets to the OBU. If you want to generate
your own messages as well (like M24), you can just pipe them through the
track model (there is an input for prior messages that will then just be
passed through and merged with the packet and message data) 

2. "In" your model. The principle is the same. But I think (1) is easier
to integrate. 

You can use the functions from RadioLib to build messages and packets. 

The "old" merge function is now a modular one. 

The model you need is called
AmsterdamUtrechtL2::AmsterdamUtrechtTrack1::Amsterdam_Utrecht_Lijn1_RBC 

It has the inputs 

Message_in: Here you can pipe your messages (from Subset 26-8) in.
Please look at RadioLib (in the BaliseLib project) for conversion
functions for all Track to Train messages from Chapter 8 

Packets_in: 

Compressed Packet data (same format as in the "old" merge function) 

Trigger_in 

Here comes your trigger, as defined in the XREF document 

Message_sent 

When you generate a message yourself, you should set this input to
"true". 

Hope that helps.- 

Mairamou 

On 2015-06-29 10:34, Alexander Probst wrote: 

> Good morning @mairamou [1]! I agree with you. 
> 
> From my point of view, I generate inside my project a trigger ID depending on LRBG and NID_LRBG. Now it would be of course fine if I could just import the UtrechtAmsterdam_oETCS.etp project into mine and call a provided operator from your side and pass this trigger value to it. As a result, your operator returns me a compressed radio message. 
> 
> For example: let us assume that you created a whole level 2 track for the radio messages similar to Story00A and called the outer operator "Track_Radio". Then e.g. the following signature would satisfy my needs: 
> 
> function Track_Radio(triggerID : int) returns (radioTrackTrainMessage : TM::CompressedRadioMessage); 
> 
> --> a trigger ID as an input and a compressed radio message as an output 
> 
> Of course you can put the whole track including the Balise messages into one operator and call it e.g. "Track". As a result, you would use more input arguments and more outputs. For me it doesn't matter as long as the operator contains an input that takes a trigger ID and returns me a compressed radio message. 
> 
> --
> Reply to this email directly or view it on GitHub [2].

-- 
Mairamou Haman Adji
Student Trainee / Embedded Systems
---------------------------
LEA RAILERGY

 

Links:
------
[1] https://github.com/mairamou
[2]
https://github.com/openETCS/modeling/issues/497#issuecomment-116523841

**Comment by aprobst** [comment]
Hello @mairamou:

Thank you for your response. Unfortunately, there seem still to exist inconsistencies and different point of views in the whole approach.

From my point of view and according to the principle of separation of concerns, the project file UtrechtAmsterdam_oETCS.etp shall only contain the track layout and a mechanism that is able to provide compressed Balise and radio based Track-to-Train messages (just an extern library). So, this project should only be included by the RBC and the EVC and not the other way round, i.e., that the track model is not allowed to know anything about the RBC or EVC.

Referring to (1): Again the principle of separation of concerns. The track model is not responsible for sending radio messages to the OBU. This is the responsibility of the RBC. The RBC only includes the UtrechtAmsterdam_oETCS.etp project, calls an operator from this project if needed and gets the compressed radio message as a result. This concept applies also to the EVC.

//------------------------------------------------------------

Reffering to (2):

The operator AmsterdamUtrechtL2::AmsterdamUtrechtTrack1::Amsterdam_Utrecht_Lijn1_RBC uses following inputs:

-	in Message_in : TM_radio_messages::M_TrackTrain_Radio_T
-	in Packets_in : Common_Types_Pkg::CompressedPackets_T
-	in Trigger_in : int
-	out Message_sent : bool
-	out Message_out : TM_radio_messages::M_TrackTrain_Radio_T
-	out Packets_out : Common_Types_Pkg::CompressedPackets_T

Now, the type TM_radio_messages::M_TrackTrain_Radio_T seems confusing and somehow incomplete. It looks like there is a mixture with the already defined common types. I had a discussion with @Farhangi  on Friday and we still do not know for which purposes these new kind of types are needed. Also, I do not understand why the RBC should provide the input for Message_in and Packets_in. From my point of view, this is not the responsibility of the RBC.

Also, I do not understand why the type TM::CompressedRadioMessage has been pushed into the background in your concept. I had several discussions with @BerndHekele and @BaseliyosJacob  and it was told that this type should always be used now.

Summing up, I personally recommend you to reduce the operator AmsterdamUtrechtL2::AmsterdamUtrechtTrack1::Amsterdam_Utrecht_Lijn1_RBC and use the following signature:

function AmsterdamUtrechtL2::AmsterdamUtrechtTrack1::Amsterdam_Utrecht_Lijn1_RBC (triggerID : int) returns (radioTrackTrainMessage : TM::CompressedRadioMessage);

Your root function just returns a compressed radio message according to the trigger ID and the RBC decides what to do with the result.

I think this will simplify the work for both sides because currently I have no idea how to proceed. 

Sorry for this long message @mairamou . But you see that this problem is not so easy like someone might think.

Best regards,

Alex

**Comment by JakobGartner** [comment]
Hello Alexander, 

let me jump in here. 
The underlying libraries (in TrackMessages and InfraLib/ RadioLib) come from me.

I will provide my answers within the text.

From an architectural point of view you are right.
The RBC module is the master of the RBC messages.

You can see the model …Lijn1_RBC like a modular replacement for the old merge radio messages operator. 
This means that you can use it it the generate your needed messages and packets.

What Mairamou tried to express is that its functionality should be placed on the output side of your system, as the outputs (the compressed packets and the messages) are then sent via the simulated air gap to the OBU.
Of course it does not know anything about the OBU nor about the RBC, with the exception of the trigger.

But it seems that there is a misconception about SCADE. You use talk about it like it would be a procedural language, which it is not.
It is a data flow language.
The Track model is, in terms of data flow, taking in information from the RBC and then generates the respective data. Whether you place it inside the RBC or after the RBC does not make any difference for SCADE (the box boundaries are irrelevant from a semantic point of view)

The track model cannot send anything alone, it needs the command from the RBC (the “trigger”) to do so. 

From a language point of view, there are no “calls” and “returns” in SCADE. All is data flow.

TM:CompressedRadioMessage did not satisfy the SRS: Many messages from Subset 26-8 contain additional variables that could not be transported using the initial data structure.
You can see TM_radio_messages::M_TrackTrain_Radio_T as a compression mechanism to transport the different messages. 
In other words, it is transparent to you.
In order to send a message, you use the function SendRadioMessage from the package  RadioLib and compose your own function using the library in MessageLibBaseline2 (you can use the track model as an example in order to include your specific 26_8 message and 26_7 packets.

Why did we do this?

In order to have a modular track model, we need operators that we can chain in a linear way and that will merge the messages and packages onto a single data flow.

Thats why the track model has a message input and a packets input (and outputs) so that the models can be concatenated as needed.
You can see the hierarchical structure of concatenated submodels in the track models in the AmsterdamUtrecht model.

Now, if the RBC wants to generate its own dynamic messages, they must be merged with the general message/packets data flow. 
Either you pipe them in at the front of the track model, or you use the aforementioned library functions in order to add them afterwards to the flow.

All this does not mean that the RBC model for Amsterdam Utrecht is not lower in hierarchy than the RBC model, in contrary. 



You can see the existing implementation as a modular way to merge an unlimited number of messages and packets onto the radio message data flow. 

I hope this makes it a bit clearer. 

Once the message is triggered, all that remains to do is to send it to the EVC.
A merge function with the balise messages will also be provided. 

What also needs to be clarified is that the CompressedPackets only carry the packets that may be part of a message. 
The messages themselves are higher in hierarchy. 

Jakob

**Comment by JakobGartner** [comment]
As an input to the discussion on data types:

The type TM::Radio_TrackTrain_Header_T is an integer version of the type Radio_Types_Pkg::Radio_TrackTrain_Header_T

This data structure combines two aspects:
- It defines a superset of the variables from 26_8 (which then, on a case- by case basis) seem to be filled out or not depending on the message to be transmitted
- it adds fields of information that are NOT relevant outside of the EVC.
1. RadioDevice
2. receivedSystemTime.

In my first attempt, I have tried to use this type but I found it not really fitting for the simulation side.
New types have therefore been defined in order to have a trackside view.

I still believe that the information should be clearly structured and modelled, that was the reason for the change. 

Overloading of variable space has been achieved by using an integer- based scheme (as for the packets)

In order to ease integration, a new conversion function will be added to translate the data type TM::Radio_TrackTrain_Header_T to and from the data type TM_radio_messages::M_TrackTrain_Radio_T

**Comment by JakobGartner** [comment]
@aprobst :

I have packaged the track model as you desired (please note that the model itself needed not to be changed)


We should however still discuss data type optimisations as far as message headers are concerned.

The RBC track model is now initialised in a way that you can't send messages/ packets through it, instead you just send the trigger and get the messages and packets in the desired formats.

As you and Bernd may have different needs, three outputs are provided.

Just terminate the one(s) that you don't need for now. I can easily delete the not needed ones then:

Compressed_Radio_Message_out: TM::CompressedRadioMessage
RadioTrackTrainHeader_out: Radio_Types_Pkg::Radio_TrackTrain_Header_T
Compressed_Packets_out: Common_Types_Pkg::CompressedPackets_T

**Comment by JakobGartner** [comment]
From Track Model point of view, this is done.

However we must discuss the data type used for the message headers, as they raise some consistency problems. I propose to use the type TM::M_TrackTrainRadio_T for transport through the EVC front-end.
TrackMessages provides the full set of extraction functions to read the messages inside the EVC, analog to what we already did for the packets.

**Comment by JakobGartner** [comment]
@aprobst 
on packet level, I have analysed. For managing your packets in your model you can continue to use the operator TM_old_versions::SEND_MergeRadioPackets.
I have updated it to comply with the new richer metadata structure. 
 (@BerndHekele : this only touches the content (data) transported, not the data type itself)


on message level, we have to find a more global solution. This will also be a discussion with @BerndHekele as it also touches the balise/radio input channel.

**Comment by BaseliyosJacob** [comment]
- [ ] Analysis
- [ ] Modelling
- [ ] Documentation
- [ ] Handover to V&V
- [ ] Approval from V&V

**Comment by BaseliyosJacob** [comment]
Handover to V&V out of scope

**Comment by BaseliyosJacob** [comment]
Dear all,

here are the JRU packets for BG 352 and 353 

![image](https://cloud.githubusercontent.com/assets/3975810/8615996/eca541ae-26f7-11e5-8175-4a4d5ba48723.png)

**Comment by BaseliyosJacob** [comment]
Seen in the JRU data packet 45 will be transmitted in BG 352 and packet 42 in BG 353

**Comment by BaseliyosJacob** [comment]
@BerndHekele and @aprobst was this the question you mentioned today in the weekly session??

**Comment by BaseliyosJacob** [comment]
- [x] Analysis
- [x] Modelling
- [ ] Documentation
- [ ] Handover to V&V
- [ ] Approval from V&V

**Comment by janWelte** [comment]
First Version of the Verification report for D3.5.3 is done and can be found under https://github.com/openETCS/validation/blob/master/Reports/VerificationReports/Trackside/. This part covers the design based on all available documentation.

Verification of the data in the SCADE model implementation will follow soon.

Verification Report: [Design Verification Report Trackside Modell] (https://github.com/openETCS/validation/blob/master/Reports/VerificationReports/Trackside/VerfRprt-DynamicTracksideModel_150821-jw-V01.pdf)

**Comment by JakobGartner** [comment]
done in large parts, US 14-15 in work by @mairamou

**Comment by JakobGartner** [comment]
correction: done. The US14-15 are covered in another issue. To be closed.
